lib/kennel/models/record.rb

```diff
@@ -129,25 +135,53 @@ def build_json
         }.compact
       end
 
+      # One of three outcomes:
+      # - throws (on non-validation error)
+      # - returns, validation_errors is not empty, and as_json is nil (invalid)
+      # - returns, validation_errors is empty, and as_json is not nil (valid)
+      def prepare
+        @validation_errors = []
+        json = nil
+
+        begin
+          json = build_json
+          (id = json.delete(:id)) && json[:id] = id
+          validate_json(json) if validate
+        rescue StandardError
+          if validation_errors.empty?
+            @validation_errors = nil
+            raise PrepareError, safe_tracking_id
+          end
+        end
+
+        @as_json = (json if validation_errors.empty?)
+      end
+
       def as_json
-        @as_json ||= begin
-                       json = build_json
-                       (id = json.delete(:id)) && json[:id] = id
-                       validate_json(json) if validate
-                       json
-                     end
+        # A courtesy to those tests that still expect as_json to perform validation and raise on error
+        prepare if @validation_errors.nil?
+        raise Kennel::ValidationError, "#{safe_tracking_id} #{@validation_errors.first}" unless validation_errors.empty?
```

Comment:
should show all since this is what users will see ?

Reply:
See my just-added comment further up.